pennylane/ops/op_math/condition.py

```diff
@@ -647,39 +665,142 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def _register_custom_staging_rule(cond_prim):
+
+    # see https://github.com/jax-ml/jax/blob/9e62994bce7c7fcbb2f6a50c9ef89526cd2c2be6/jax/_src/lax/lax.py#L3538
+    # and https://github.com/jax-ml/jax/blob/9e62994bce7c7fcbb2f6a50c9ef89526cd2c2be6/jax/_src/lax/lax.py#L208
+    # for reference to how jax is handling staging rules for dynamic shapes in v0.4.28
+    # see also capture/intro_to_dynamic_shapes.md
+
+    import jax  # pylint: disable=import-outside-toplevel
+    from jax._src.interpreters import partial_eval as pe  # pylint: disable=import-outside-toplevel
+
+    def _tracer_and_outvar(
+        jaxpr_trace: pe.DynamicJaxprTrace,
+        outvar: jax.core.Var,
+        env: dict[jax.core.Var, jax.core.Var],
+    ) -> tuple[pe.DynamicJaxprTracer, jax.core.Var]:
+        """
+        Create a new tracer and returned var from the true branch outvar
+        returned vars are cached in env for use in future shapes
+        """
+        if not hasattr(outvar.aval, "shape"):
+            out_tracer = pe.DynamicJaxprTracer(jaxpr_trace, outvar.aval)
+            return out_tracer, jaxpr_trace.makevar(out_tracer)
+        new_shape = [s if isinstance(s, int) else env[s] for s in outvar.aval.shape]
+        new_aval = jax.core.DShapedArray(tuple(new_shape), outvar.aval.dtype)
+        out_tracer = pe.DynamicJaxprTracer(jaxpr_trace, new_aval)
+        new_var = jaxpr_trace.makevar(out_tracer)
+
+        if not isinstance(outvar, jax.core.Literal):
+            env[outvar] = new_var
+        return out_tracer, new_var
+
+    def custom_staging_rule(
+        jaxpr_trace: pe.DynamicJaxprTrace, *tracers: pe.DynamicJaxprTracer, **params
+    ) -> Sequence[pe.DynamicJaxprTracer] | pe.DynamicJaxprTracer:

```

Comment:
```suggestion
    ) -> Union[Sequence[pe.DynamicJaxprTracer], pe.DynamicJaxprTracer]:
```
Sphinx seems unhappy: https://github.com/PennyLaneAI/pennylane/actions/runs/13637870947/job/38120911554?pr=6888.
Probably this change works?

Reply:
Looks like even though we dropped support for 3.9, Sphinx is still sticking to it....